numpy/random/bit_generator.pyx

```diff
@@ -0,0 +1,458 @@
+        if self.program_entropy is None:
+            # We *could* make `coerce_to_uint32_array(None)` handle this case,
+            # but that would make it easier to misuse, a la
+            # `coerce_to_uint32_array([None, 12345])`
+            program_entropy = np.array([], dtype=np.uint32)
+        else:
+            program_entropy = coerce_to_uint32_array(self.program_entropy)
+        spawn_entropy = coerce_to_uint32_array(self.spawn_key)
+        entropy_array = np.concatenate([run_entropy, program_entropy,
+                                        spawn_entropy])
+        return entropy_array
+
+    @np.errstate(over='ignore')
+    def generate_state(self, n_words, dtype=np.uint32):
+        """ Return the requested number of words for PRNG seeding.
+
+        A BitGenerator should call this method in its constructor with
+        an appropriate `n_words` parameter to properly seed itself.
+
+        Parameters
+        ----------
+        n_words : int
+        dtype : np.uint32 or np.uint64, optional
+            The size of each word. This should only be either `uint32` or
+            `uint64`. Strings (`'uint32'`, `'uint64'`) are fine. Note that
+            requesting `uint64` will draw twice as many bits as `uint32` for
+            the same `n_words`. This is a convenience for `BitGenerator`s that
+            express their states as `uint64` arrays.
+        Returns
+        -------
+        state : uint32 or uint64 array, shape=(n_words,)
+        """
+        hash_const = INIT_B
+        out_dtype = np.dtype(dtype)
+        if out_dtype == np.dtype(np.uint32):
+            pass
+        elif out_dtype == np.dtype(np.uint64):
+            n_words *= 2
+        else:
+            raise ValueError("only support uint32 or uint64")
+        state = np.zeros(n_words, dtype=np.uint32)
+        src_cycle = cycle(self.pool)
+        for i_dst in range(n_words):
+            data_val = next(src_cycle)
+            data_val ^= hash_const
+            hash_const *= MULT_B
+            data_val *= hash_const
+            data_val ^= data_val >> XSHIFT
+            state[i_dst] = data_val & MASK32
+        if out_dtype == np.dtype(np.uint64):
+            state = state.view(np.uint64)
+        return state
+
+    def spawn(self, n_children):
+        """ Spawn a number of child `SeedSequence`s by extending the
+        `spawn_key`.
+        Parameters
+        ----------
+        n_children : int
+        Returns
+        -------
+        seqs : list of `SeedSequence`s
+        """
+        seqs = []
+        for i in range(self.n_children_spawned,
+                       self.n_children_spawned + n_children):
+            seqs.append(SeedSequence(
+                self.entropy,
+                program_entropy=self.program_entropy,
+                spawn_key=self.spawn_key + (i,),
+                pool_size=self.pool_size,
+            ))
+        return seqs
+
+cdef class BitGenerator():
+    """
+    BitGenerator(seed=None)
```

Comment:
Should BItGenerator be private?  It is not useful to instantize.

Reply:
As a user of randomgen, I would request that the base class for `BitGenerator` be public. Having a public base class rather simplifies [PEP 484](https://www.python.org/dev/peps/pep-0484/) static typing and it can be used in `isinstance` tests, etc.

My use case is an API for a Monte Carlo simulation package I'm working on that accepts a seed that is either an `int` or a `Generator`. Some internal code needs to pick a BitGenerator, and, to keep Mypy happy, I had to create a [`typing.Union`](https://docs.python.org/3/library/typing.html#typing.Union) of the BitGenerators I wanted to allow. This is admittedly a minor annoyance, but one that would not have been an issue at all if a base class were available in randomgen.

**Proposal**: Making `BitGenerator`'s methods be [abstract](https://docs.python.org/3/library/abc.html#abc.abstractmethod) would solve your (legitimate!) API concern than `BitGenerator` is not useful to instantiate. It would also have the benefit of [`inspect.isabstract(BitGenerator)`](https://docs.python.org/3/library/inspect.html#inspect.isabstract) returning true. A final benefit is that you could use the docstrings of the abstract methods to document the internal API.

Reply:
I played with different strategies and googled for a bit, it seems cython does not support abstract base classes. Please correct me if I am wrong. 

I think there is value in exposing the base class for documentation and for re-implementors of other BitGenerators. I added getter/setter for `state`that raises a `NotImplementedError` as a best effort.